lib/src/test/kotlin/graphql/nadel/engine/NadelIncrementalResultSupportTest.kt

```diff
@@ -0,0 +1,214 @@
+        val channel = Channel<DelayedIncrementalPartialResult>(UNLIMITED)
+
+        val subject = NadelIncrementalResultSupport(channel)
+        val firstLock = CompletableDeferred<Boolean>()
+        val secondLock = CompletableDeferred<Boolean>()
+
+        // When
+        subject.defer {
+            // Wait until test tells us to continue
+            secondLock.join()
+
+            DelayedIncrementalPartialResultImpl.newIncrementalExecutionResult()
+                .incrementalItems(
+                    listOf(
+                        DeferPayload.newDeferredItem()
+                            .data("Hello world")
+                            .path(listOf("echo"))
+                            .build()
+                    )
+                )
+                .hasNext(false)
+                .build()
+        }
+
+        subject.defer {
+            // Wait until test tells us to continue
+            firstLock.join()
+
+            DelayedIncrementalPartialResultImpl.newIncrementalExecutionResult()
+                .incrementalItems(emptyList())
+                .hasNext(false)
+                .build()
+        }
+
+        // Then
+        firstLock.complete(true)
+        val firstItem = channel.receive()
+        assertTrue(firstItem.incremental?.isEmpty() == true)
+        assertTrue(firstItem.hasNext())
+
+        secondLock.complete(true)
+        val secondItem = channel.receive()
+        assertTrue(secondItem !== firstItem)
+        assertTrue(secondItem.incremental?.isNotEmpty() == true)
+        assertFalse(secondItem.hasNext())
+    }
+
+    @Test
+    fun `channel still closes if the last defer job fails`() {
+        var testCompleted = false
+
+        try {
+            runTest {
+                val channel = Channel<DelayedIncrementalPartialResult>(UNLIMITED)
+
+                val subject = NadelIncrementalResultSupport(channel)
+                val lock = CompletableDeferred<Boolean>()
+
+                assertFalse(channel.isClosedForSend)
+
+                // When
+                subject.defer {
+                    // Wait until test tells us to continue
+                    lock.join()
+                    throw UnsupportedOperationException("Hello")
+                }
+                subject.onInitialResultComplete()
+
+                // Then
+                lock.complete(true)
+                // Can only get last element once the channel is closed
+                assertTrue(channel.consumeAsFlow().lastOrNull() == null)
+
+                // Must be at end of runTest
+                testCompleted = true
+            }
+        } catch (e: UnsupportedOperationException) {
+            assertTrue(e.message == "Hello")
+        }
+
+        // Coroutines code will throw uncaught exceptions even though we threw it deliberately
+        // Just ensure that we actually ran all the asserts and the tests is fine
+        assertTrue(testCompleted)
```

Comment:
Amazing test coverage @gnawf . Can we add some tests for the `Flow` version of the `defer` function though?